rpc,server: use node certs for shared-process tenant RPCs

knz opened the pull request:
TLDR: This PR makes it possible to use regular node certs with shared-process multi-tenancy (i.e. skip creating client tenant certs). I think this will help @aadityasondhi 's work on `debug zip`.

Fixes #96215.
Epic: CRDB-23559

The PR also contains a few cleanup commits; it's a short change away from fixing  #87141 and #77173, which I intend to do in a separate PR.

knz committed:
rpc: mandate TenantID in ContextOptions

This will protect against programming errors.

Release note: None

knz committed:
server: add a missing comment

Release note: None

knz committed:
rpc,security: simplify error handling

The security package was using pre-go 1.13 error types. This was
crufty and also caused the error payloads to be unduly redacted in
logs. This patch fixes it.

Release note: None

knz committed:
security: extract `GetNodeClientTLSConfig()` into a separate function

Prior to this patch, `GetClientTLSConfig()` was used both
to obtain a TLS config suitable for a standalone/CLI RPC client
and also for the TLS config inside a KV server.

This was confusing and error prone. We really have two cases
and they are quite different:

- a standalone RPC client (e.g. a CLI tool) connecting
  to a remote RPC server, under some non-node identity.
  This is what `GetClientTLSConfig()` is for.

- a RPC server for a KV node that's also RPC client of another KV
  node. In that case, this server can know it has server (node) certs to
  work with. This is the new `GetNodeClientTLSConfig()`

- a RPC server for a secondary tenant server that's also RPC client to
  something else (either another server for the secondary tenant, or a
  KV node). This is what `GetTenantTLSConfig()` is for.

Release note: None

knz committed:
rpc,server: use node certs for shared-process tenant RPCs

When a tenant server running in the same process as a KV node makes an
outgoing RPC, it can use the same node certs as the KV node.
This has two advantages:

- it is simpler to orchestrate; this way we do not need to provision
  the tenant client certs manually;
- it paves the road for using a single RPC server for all tenants
  on a server (future work).

Release note: None

knz committed:
cli/democluster: only generate tenant certs if necessary

Now that the RPC code is able to perform tenant-to-tenant RPCs without
tenant client certs, we can demonstrate this by disabling the
auto-generation of a tenant client cert when shared-process servers
are used.

Release note: None

knz committed:
cli: de-flake the zip tests

Goroutine / heap / profile dump generation is non-deterministic.
The reference test output should not include them.

Release note: None

knz committed:
rpc: double check client username is set with ClientOnly

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96553)
<!-- Reviewable:end -->

knz committed:
rpc,security: change `rpc.MakeSecurityContext` to `NewSecurityContext`

This constructor makes an object that contains a mutex lock. We have
linters that find it illegal to copy an object that contains a lock.
So this patch fixes that by ensuring that struct is instantiated on
the heap.

Release note: None

stevendanna reviewed (approved):
Thanks for pushing this forward.

knz commented:
thank you!

bors r=stevendanna

knz commented:
bors r=stevendanna
